gunrock/oprtr/advance/advance_base.cuh

```diff
@@ -143,49 +113,34 @@ template <OprtrFlag FLAG, typename GraphT, typename FrontierInT,
 cudaError_t ComputeOutputLength(const GraphT graph,
                                 const FrontierInT *frontier_in,
                                 ParameterT &parameters) {
-  // Load Load Balanced Kernel
-  // Get Rowoffsets
-  // Use scan to compute edge_offsets for each vertex in the frontier
-  // Use sorted sort to compute partition bound for each work-chunk
-  // load edge-expand-partitioned kernel
 
   typedef typename GraphT::SizeT SizeT;
   typedef typename FrontierInT::ValueT InKeyT;
   cudaError_t retval = cudaSuccess;
   if (parameters.frontier->queue_length == 0) {
-    // printf("setting output_length to 0");
     oprtr::Set_Kernel<<<1, 1, 0, parameters.stream>>>(
         parameters.frontier->output_length.GetPointer(util::DEVICE), 0, 1);
     return retval;
   }
 
-  // util::PrintMsg("output_offsets.size() = "
-  //    + std::to_string(parameters.frontier -> output_offsets.GetSize())
-  //    + ", queue_length = "
-  //    + std::to_string(parameters.frontier -> queue_length)
-  //    + ", queue_size = "
-  //    + std::to_string(frontier_in == NULL ? 0 : frontier_in -> GetSize()));
-
   int block_size = 512;
   SizeT num_blocks = parameters.frontier->queue_length / block_size + 1;
   if (num_blocks > 80 * 4) num_blocks = 80 * 4;
   GetEdgeCounts<<<num_blocks, block_size, 0, parameters.stream>>>(
       graph,
       (frontier_in == NULL) ? (InKeyT *)NULL
                             : frontier_in->GetPointer(util::DEVICE),
-      parameters.frontier->queue_length,  // TODO: +1?
+      parameters.frontier->queue_length,
       parameters.frontier->output_offsets.GetPointer(util::DEVICE), FLAG);
-  // util::DisplayDeviceResults(partitioned_scanned_edges,
-  // frontier_attribute->queue_length);
-
-  mgpu::Scan<mgpu::MgpuScanTypeInc>(
-      parameters.frontier->output_offsets.GetPointer(util::DEVICE),
-      parameters.frontier->queue_length,  // TODO: +1?
-      (SizeT)0, mgpu::plus<SizeT>(),
-      parameters.frontier->output_length.GetPointer(util::DEVICE),
-      (SizeT *)NULL,
-      parameters.frontier->output_offsets.GetPointer(util::DEVICE),
-      parameters.context[0]);
+
+  util::cubInclusiveSum(parameters.frontier->cub_temp_space,
+                        /*d_in=*/parameters.frontier->output_offsets,
+                        /*d_out=*/parameters.frontier->output_offsets,
+                        /*num_items=*/parameters.frontier->queue_length);
+  GUARD_CU2(
+          cudaMemcpy(parameters.frontier->output_length.GetPointer(util::DEVICE),
```

Comment:
There should be an Array1D utility for this copy, I believe you should just be able to do `.Move` after setting the pointer, and the move will be from `util::DEVICE` to `util::DEVICE`. We'd like to avoid using direct cuda APIs in the code without the wrapper.

Reply:
Yeah I would really like to avoid. But Move() only works on same Array1D object? I checked there is no such Move() function. Or did I miss it? We can of course add one later to provide Move() between Array1D objects. I will add a TODO.

Reply:
No, it could be used across any pointer. At least we do that for all Extract functions in problem by setting the pointer of an Array1D to a new location, and then initiating a move on it. Let me look it up to make sure.

Reply:
It may be missing `source == DEVICE && target == DEVICE`, I didn't get a chance to properly go over it: https://github.com/gunrock/gunrock/blob/master/gunrock/util/array_utils.cuh#L771-L884

Reply:
Or, like you said, we could always write it and have it a `TODO`.

Reply:
Thanks for the link! In this context I actually want to do a d2d memcpy between two Array1D objects. I think a simple function like this: https://github.com/moderngpu/moderngpu/blob/master/src/moderngpu/memory.hxx#L38 would do the trick. Our Move() is missing a pointer to other Array1D object in its interface.
Let me put a TODO and I can add this in later commit.
Another option is to directly send in the device pointer to total length in scan. That would save an explicit memcpy outside the kernel but alas, we need to change cub implementation.